src/Integration/VsSessionHost.cs

```diff
@@ -265,19 +262,21 @@ private BoundSonarQubeProject SafeReadBindingInformation()
         #region IServiceProvider
         private void RegisterLocalServices()
         {
-            // Use Lazy<object> to avoid creating instances needlessly
-            this.localServices.Add(typeof(ISolutionRuleSetsInformationProvider), new Lazy<ILocalService>(() => new SolutionRuleSetsInformationProvider(this)));
-            this.localServices.Add(typeof(IRuleSetSerializer), new Lazy<ILocalService>(() => new RuleSetSerializer()));
-            this.localServices.Add(typeof(ISolutionBinding), new Lazy<ILocalService>(() => new SolutionBinding(this)));
-            this.localServices.Add(typeof(IProjectSystemHelper), new Lazy<ILocalService>(() => new ProjectSystemHelper(this)));
-            this.localServices.Add(typeof(IConflictsManager), new Lazy<ILocalService>(() => new ConflictsManager(this)));
-            this.localServices.Add(typeof(IRuleSetInspector), new Lazy<ILocalService>(() => new RuleSetInspector(this)));
-            this.localServices.Add(typeof(IRuleSetConflictsController), new Lazy<ILocalService>(() => new RuleSetConflictsController(this)));
-            this.localServices.Add(typeof(IProjectSystemFilter), new Lazy<ILocalService>(() => new ProjectSystemFilter(this)));
+            IServiceContainer container = this;
 
+            container.AddService(typeof(ISolutionRuleSetsInformationProvider), (c, t) => new SolutionRuleSetsInformationProvider(this));
+            container.AddService(typeof(IRuleSetSerializer), (c, t) => new RuleSetSerializer());
+            container.AddService(typeof(ISolutionBinding), (c, t) => new SolutionBinding(this));
+            container.AddService(typeof(IProjectSystemHelper), (c, t) => new ProjectSystemHelper(this));
+            container.AddService(typeof(IConflictsManager), (c, t) => new ConflictsManager(this));
+            container.AddService(typeof(IRuleSetInspector), (c, t) => new RuleSetInspector(this));
+            container.AddService(typeof(IRuleSetConflictsController), (c, t) => new RuleSetConflictsController(this));
+            container.AddService(typeof(IProjectSystemFilter), (c, t) => new ProjectSystemFilter(this));
+
+            // Use Lazy<object> to avoid creating instances needlessly
             var sccFs = new Lazy<ILocalService>(() => new SourceControlledFileSystem(this));
-            this.localServices.Add(typeof(ISourceControlledFileSystem), sccFs);
-            this.localServices.Add(typeof(IFileSystem), sccFs);
+            container.AddService(typeof(ISourceControlledFileSystem), (c, t) => sccFs.Value);
+            container.AddService(typeof(IFileSystem), (c, t) => sccFs.Value);
```

Comment:
It would useful to call out that the same instance must be returned for _ISourceControllerFileSystem_ and _IFileSystem_.
Alternatively, you get rid of the need for the additional comment and the _lazy_ by doing something like
```
container.AddService(typeof(ISourceControlledFileSystem), (c, t) => new SourceControlledFileSystem(this));
container.AddService(typeof(IFileSystem), (c, t) => (IFileSystem)c.GetService(typeof(IFileSystem)));
```